Implement proc macro feature.

daxpedda opened the pull request:
Implement proc macro alternative to `rusty_fork_test!`.
Theres a bit of documentation missing and maybe more unit tests could help, I'm willing to do it, but would like some direction.

Fixes #5 and #9.

daxpedda commented on Cargo.toml:
```diff
@@ -19,14 +19,19 @@ fork-like interface.
 [dependencies]
 fnv = "1.0"
 quick-error = "1.2"
+rusty-fork-macro = { path = "rusty-fork-macro", optional = true }
```
Should this even be optional?

daxpedda commented on rusty-fork-macro/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+version = "0.1.0"
+authors = ["Jason Lingle"]
+license = "MIT/Apache-2.0"
+readme = "README.md"
```
The actual `README.md` file here is missing, I would propose just to link back to the root one: `../README.md`.

daxpedda replied:
I linked back to the original `README.md`.

daxpedda commented on rusty-fork-macro/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+name = "rusty-fork-macro"
+version = "0.1.0"
+authors = ["Jason Lingle"]
+license = "MIT/Apache-2.0"
```
I am unsure how this is usually handled, but as far as I understand it we should make seperate LICENSE file for this crate or alternatively combine the `LICENSE-APACHE` and `LICENSE-MIT` files into one and link to it via `license-file`.

daxpedda replied:
I compared it to official rust repositories, the all just fill the `license` field cooreclty, thats it.

daxpedda commented on rusty-fork-macro/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+edition = "2018"
+
+description = """
+Proc-macro library for rusty-fork.
```
Any input on this would be appreciated.

daxpedda replied:
I believe this is fine like that.

daxpedda commented on rusty-fork-macro/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+
+[dev-dependencies]
+rusty-fork = { path = "../", features = ["macro", "timeout"] }
+tokio = { version = "0.2", features = ["rt-threaded", "macros"] }
```
I used tokio to test the `async` example, but any other crate would do too.

daxpedda replied:
`tokio` is currently the most used general async library out there, I think this is fine.

daxpedda commented on rusty-fork-macro/src/lib.rs:
````diff
@@ -0,0 +1,210 @@
+///     use rusty_fork::test_fork;
+///
+///     # /*
+///     #[test_fork]
+///     # */
+///     # pub
+///     fn my_test() {
+///         assert_eq!(2, 1 + 1);
+///     }
+///
+///     // more tests...
+/// }
+/// #
+/// # fn main() { test::my_test(); }
+/// ```
+///
+/// Each test will be run in its own process. If the subprocess exits
+/// unsuccessfully for any reason, including due to signals, the test fails.
+///
+/// It is also possible to specify a timeout which is applied to all tests in
+/// the block, like so:
+///
+/// ```
+/// use rusty_fork::test_fork;
+///
+/// # /*
+/// #[test_fork(timeout_ms = 1000)]
+/// # */
+/// fn my_test() {
+///     do_some_expensive_computation();
+/// }
+/// # fn do_some_expensive_computation() { }
+/// # fn main() { my_test(); }
+/// ```
+///
+/// If any individual test takes more than the given timeout, the child is
+/// terminated and the test panics.
+///
+/// Using the timeout feature requires the `timeout` feature for this crate to
+/// be enabled (which it is by default).
````
Any directions on the documentation here would be appreciated.

daxpedda replied:
I went through this, seems good enough to me.

daxpedda commented on rusty-fork-macro/src/lib.rs:
```diff
@@ -0,0 +1,210 @@
+                                timeout = int.to_token_stream();
+                            }
+                        }
+                        "crate" => {
+                            if let Lit::Str(str) = name_value.lit {
+                                crate_name = str.to_token_stream();
+                            }
+                        }
```
This feature is undocumented as above.

daxpedda replied:
I documented this now.

debnil commented:
@daxpedda: I'd like to use `rust-fork` to test an `async` function using `tokio`, and this PR looks perfect for that.

Are there any plans to merge this? If you need additional community help to finish this, I'm happy to contribute.

daxpedda commented:
I will take a look at it, if anybody who his responsible for the project can answer some of my questions above that would be splendid.
The PR is done, my questions above, except documentation, are purely organizational, but otherwise the PR is ready to be merged.

daxpedda commented:
So looking through hit again, I realized that this PR is not as clean as I would like it.
It tries to solve two problems at once, that is implement a proc macro, and return a `Result`. I think this PR would be better served if #10 would be merged first.
#10 is already finished and ready to be merged as far as I am concerned.

daxpedda commented:
I cleaned everything up now:
- added documentation
- added more tests
- updated tokio
- cleaned up macro code to rely on #10

This is now only blocked on #10, which again, is ready to be merged as far as I am concerned.

debnil commented:
Wow - thanks so much for the fast turnaround and excellent work!

@AltSysrq: this pull request, along with #10, would extend support for async tests. Merging both would be incredibly helpful. Thank you so much for your great work!
